ls1intum/Artemis
# `Quiz exercises`: Short Answer Number Spot

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



[#5459](https://github.com/ls1intum/Artemis/issues/5459)

### Description


- Added type to Short Answer Spot
- Modified Short Answer Submission
- Added Add Number Spot button in Short Answer Question configuration
- Added Validation of Number Spot Solution
- Modified the input type of Short Answer Spot during Quiz Exercise

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course

1. Log in to Artemis as an Instructor
2. Navigate to Course Management
3. Select the Course
4. Create a Quiz Exercise
5. Add Short Answer Question
6. Clear the Question Template
7. Click **Add Number Spot**
8. The option of number spot is in inclusive range format (e.g. `[0.5|10]` to set the solution between 0.5 and 10 inclusive)
9. Input other text that contains a number (e.g. `Test 1.23456`)
10. Click Visual Tab
11. Highlight the text (i.e. `Test`) with the cursor
12. Click **Add Number Spot** (This should not create a new spot because the highlighted text is not a number)
13. Highlight the number (i.e. `1.23456`) with the cursor
14. Click **Add Number Spot** (This will create a new number spot whose answer is the highlighted number `[1.23456|1.23456]`)
15. Save the Quiz Exercise
16. Click **Preview Solution** (Verify that the solutions are displayed correctly)
17. Start the Quiz Exercise
18. Log out and Log in as a Student
19. Start the Quiz Exercise
20. Input to the number spot should be limited to a number only
21. Submit (Verify that the grade is calculated correctly)

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| ShortAnswerSubmittedText.java | 87% | 95% |
| ShortAnswerSpot.java | 75% | 98% |
| quiz-exercise-detail.component.ts | 80.28% | 95.62% |
| quiz-exercise-validation.directive.ts | 91.01% | 97.33 |
| short-answer-question-edit.component.ts | 73.21% | 94.24% |
| short-answer-question.component.ts | 80.59% | 92.55% |
| short-answer-question-util.service.ts | 81.35% | 95.58% |

### Screenshots



<img width="1365" alt="Screen Shot 2022-11-24 at 2 12 35 PM" src="https://user-images.githubusercontent.com/10885503/203793238-3f4c4504-2f0e-4366-a78b-ba8a8b4f65fb.png">
<img width="1350" alt="Screen Shot 2022-11-24 at 2 13 28 PM" src="https://user-images.githubusercontent.com/10885503/203793248-9557e997-8cc3-4476-b71e-05c636e6ed8f.png">
<img width="1202" alt="Screen Shot 2022-11-24 at 2 14 19 PM" src="https://user-images.githubusercontent.com/10885503/203793254-d67f49ee-8ff9-43ff-a8a4-9d501be54364.png">
